app/src/main/java/org/simple/clinic/scanid/ui/ShortCodeSpanWatcher.kt

```diff
@@ -0,0 +1,56 @@
+package org.simple.clinic.scanid.ui
+
+import android.graphics.Canvas
+import android.graphics.Paint
+import android.text.Editable
+import android.text.Spannable
+import android.text.style.ReplacementSpan
+import kotlin.math.roundToInt
+
+private const val CHARS_REQUIRED_TO_INSERT_SPACE = 4
+private const val INSERT_SPACE_AT_INDEX = 2
+
+class ShortCodeSpanWatcher : TextWatcherAdapter() {
+    private var activeSpan: AddTrailingSpaceSpan? = null
+
+    override fun afterTextChanged(editable: Editable) {
+        with(editable) {
+            if (length >= CHARS_REQUIRED_TO_INSERT_SPACE) {
+                removeSpan(activeSpan)
+                activeSpan = AddTrailingSpaceSpan()
+                setSpan(activeSpan, INSERT_SPACE_AT_INDEX, INSERT_SPACE_AT_INDEX + 1, Spannable.SPAN_INCLUSIVE_EXCLUSIVE)
+            } else {
+                removeSpan(activeSpan)
+                activeSpan = null
+            }
+        }
+    }
+}
+
+private class AddTrailingSpaceSpan : ReplacementSpan() {
+    override fun getSize(
+        paint: Paint,
+        text: CharSequence?,
+        start: Int,
+        end: Int,
+        fm: Paint.FontMetricsInt?
+    ): Int {
+        text ?: return 0
+        return paint.measureText("${text.substring(start, end)} ").roundToInt()
```

Comment:
Can we make this a non-breaking space? Check out the file `org.simple.clinic.util.Unicode` where we already have the character defined.

Reply:
We don't need a non-breaking space here, because we are only measuring the 'width' of a space character. The space character itself doesn't get rendered on the screen. We use it to decide how much area on the canvas we need to render the character. Concerns?

Reply:
Makes sense.